Automated Smoketest: :x:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: b0116ce51776d9da538226c06a04227c5c423e05
Build: success
Build: failed!
Number of Error(s): 3
src/Graph7Widget.ts(76,39): error TS2339: Property fetchScopeGraphs does not exist on type Workunit.
ECL Watch:  Rebuilding Site
Build: failed!
Number of Error(s): 3
src/Graph7Widget.ts(76,39): error TS2339: Property fetchScopeGraphs does not exist on type Workunit.
CPack Error: Error when generating package: hpccsystems-platform
make: *** [package] Error 1
CMakeFiles/CMakeOutput.log copied into ../CMakeOutput-2018-09-25_04-29-17.log
CMakeFiles/CMakeError.log copied into ../CMakeError-2018-09-25_04-29-17.log
Finish at: 2018-09-25_04-33-21
time logs started
Prep time   :    25 sec (00:00:25)
Build time  :   184 sec (00:03:04)
Package time:    28 sec (00:00:28)
Install time:     0 sec (00:00:00)
Start time  :     0 sec (00:00:00)
Test time   :     0 sec (00:00:00)
Stop time   :     0 sec (00:00:00)
Summary     :   237 sec (00:03:57)
time logs end

[Build error:](https://gist.github.com/26764ecaa7650f30a91ae0b5f92339d1)
Build log [Build log](https://gist.github.com/HPCCSmoketest/26764ecaa7650f30a91ae0b5f92339d1#file-result-18-09-25-04-29-07-log)

--- a/esp/src/eclwatch/ECLPlaygroundWidget.js
+++ b/esp/src/eclwatch/ECLPlaygroundWidget.js
@@ -15,7 +15,7 @@ define([
     "hpcc/_Widget",
     "hpcc/ECLSourceWidget",
     "hpcc/TargetSelectWidget",
-    "hpcc/JSGraphWidget",
+    "src/Graph7Widget",
     "hpcc/ECLPlaygroundResultsWidget",
     "src/ESPWorkunit",
     "src/ESPQuery",
@@ -27,14 +27,14 @@ define([
 
 ], function (declare, lang, i18n, nlsHPCC, xhr, dom, query,
     BorderContainer, TabContainer, ContentPane, registry,
-    _Widget, EclSourceWidget, TargetSelectWidget, JSGraphWidget, ResultsWidget, ESPWorkunit, ESPQuery, Utility,
+    _Widget, EclSourceWidget, TargetSelectWidget, Graph7Widget, ResultsWidget, ESPWorkunit, ESPQuery, Utility,
     template) {
         return declare("ECLPlaygroundWidget", [_Widget], {
             templateString: template,
             baseClass: "ECLPlaygroundWidget",
             i18n: nlsHPCC,
 
-            graphType: "JSGraphWidget",
+            graphType: "Graph7Widget",
             wu: null,
             editorControl: null,
             graphControl: null,
@@ -116,10 +116,7 @@ define([
                     this.watchWU();
                 } else {
                     this.initSamples();
-                    this.graphControl.watchSelect(this.sampleSelectWidget);
                 }
-                this.graphControl.watchSplitter(this.borderContainer.getSplitter("right"));
-                this.graphControl.watchSplitter(this.borderContainer.getSplitter("bottom"));
 
                 this.graphControl.onDoubleClick = function (globalID, keyState) {
                     if (keyState && context.main.KeyState_Shift) {
@@ -150,9 +147,7 @@ define([
             initGraph: function () {
                 var context = this;
                 this.graphControl = registry.byId(this.id + "GraphControl");
-                this.graphControl.showNextPrevious(false);
-                this.graphControl.showDistance(false);
-                this.graphControl.showSyncSelection(false);
+                this.graphControl.init({});
                 this.graphControl.onSelectionChanged = function (items) {
                     context.editorControl.clearHighlightLines();
                     for (var i = 0; i < items.length; ++i) {
@@ -267,17 +262,10 @@ define([
             },
 
             displayGraphs: function (graphs) {
-                var fetchedCount = 0;
-                for (var i = 0; i < graphs.length; ++i) {
-                    var context = this;
-                    this.wu.fetchGraphXgmml(i, null, function (xgmml) {
-                        ++fetchedCount;
-                        context.graphControl.loadXGMML(xgmml, fetchedCount > 1);
-                        if (fetchedCount === graphs.length) {
-                            context.graphControl.startLayout("dot");
-                        }
-                    });
-                }
+                this.graphControl.params = {
+                    Wuid: this.wu.Wuid
+                };
+                this.graphControl.doInit(this.wu.Wuid);
             },
 
             displayAll: function (workunit) {

--- a/esp/src/eclwatch/templates/ECLPlaygroundWidget.html
+++ b/esp/src/eclwatch/templates/ECLPlaygroundWidget.html
@@ -3,7 +3,7 @@
         <div id="${id}TopPane" class="topPanel" data-dojo-props="region: 'top'" data-dojo-type="dijit.Toolbar">
             <span id="${id}Title" style="font-weight: bold; font-size: x-large">${i18n.title_ECLPlayground}</span>
             <div id="${id}NewPage" class="right" data-dojo-attach-event="onClick:_onNewPage" data-dojo-props="iconClass:'iconNewPage', showLabel:false" data-dojo-type="dijit.form.Button">${i18n.OpenInNewPage}</div>
-            <div style="float: right; display: inline-block;">
+            <div style="float: right; display: block;">
                 <label id="${id}SampleSelectLabel" for="${id}SampleSelect">${i18n.Sample}:</label>
                 <div id="${id}SampleSelect" data-dojo-type="TargetSelectWidget">
                 </div>
@@ -32,7 +32,7 @@
                 <div id="${id}StackController" data-dojo-props="containerId:'${id}StackContainer'" data-dojo-type="dijit.layout.StackController"></div>
                 <span data-dojo-type="dijit.ToolbarSeparator"></span>
             </div>
-            <div style="float: right; display: inline-block; vertical-align: middle">
+            <div style="float: right; vertical-align: middle">
                 <a id="${id}State" target="_blank">...</a>
             </div>
         </div>

--- a/esp/src/eclwatch/templates/Graph7Widget.html
+++ b/esp/src/eclwatch/templates/Graph7Widget.html
@@ -1,6 +1,21 @@
 <div class="${baseClass}">
-    <div id="${id}BorderContainer" class="${baseClass}BorderContainer" style="width: 100%; height: 100%; padding: 0px; overflow: hidden" data-dojo-props="splitter: false, gutters:false" data-dojo-type="dijit.layout.BorderContainer">
-        <div id="${id}GraphContentPane" class="${baseClass}GraphContentPane" style="padding: 0px; overflow: hidden" data-dojo-props="region: 'center'" data-dojo-type="dijit.layout.ContentPane">
+    <div id="${id}MainBorderContainer" class="${baseClass}BorderContainer" style="width: 100%; height: 100%; padding: 0px; overflow: hidden" data-dojo-props="splitter: false, gutters:false" data-dojo-type="dijit.layout.BorderContainer">
+        <div id="${id}GraphToolbarContentPane" class="${baseClass}ToolbarContentPane" style="padding: 0px; overflow: hidden" data-dojo-props="region: 'top'" data-dojo-type="dijit.layout.ContentPane">
+            <div id="${id}GraphToolbar" class="topPanel dijit dijitToolbar" role="toolbar">
+                <div data-dojo-attach-event="onClick:_onGraphRefresh" data-dojo-props="iconClass:'fa fa-refresh', showLabel:false" data-dojo-type="dijit.form.Button">${i18n.Refresh}</div>
+                <span data-dojo-type="dijit.ToolbarSeparator"></span>
+                <div data-dojo-attach-event="onClick:_onPartial" data-dojo-props="iconClass:'fa fa-window-restore', showLabel:false" data-dojo-type="dijit.form.Button">${i18n.CollapseAll}</div>
+                <div data-dojo-attach-event="onClick:_onMax" data-dojo-props="iconClass:'fa fa-window-maximize', showLabel:false" data-dojo-type="dijit.form.Button">${i18n.Restore}</div>
+                <span data-dojo-type="dijit.ToolbarSeparator"></span>
+                <div data-dojo-attach-event="onClick:_onZoomToFit" data-dojo-props="iconClass:'fa fa-arrows-alt', showLabel:false" data-dojo-type="dijit.form.Button">${i18n.ZoomAll}</div>
+                <div data-dojo-attach-event="onClick:_onZoomToWidth" data-dojo-props="iconClass:'fa fa-arrows-h', showLabel:false" data-dojo-type="dijit.form.Button">${i18n.ZoomWidth}</div>
+                <div data-dojo-attach-event="onClick:_onZoomToPlus" data-dojo-props="iconClass:'fa fa-plus', showLabel:false" data-dojo-type="dijit.form.Button">${i18n.ZoomPlus}</div>
+                <div data-dojo-attach-event="onClick:_onZoomToMinus" data-dojo-props="iconClass:'fa fa-minus', showLabel:false" data-dojo-type="dijit.form.Button">${i18n.ZoomMinus}</div>
+                <span data-dojo-type="dijit.ToolbarSeparator"></span>
+                <span id="${id}GraphStatus" ></span>
+            </div>
+        </div>
+        <div id="${id}MainGraphWidget" style="overflow: hidden" data-dojo-props="region: 'center'" data-dojo-type="dijit.layout.ContentPane">
         </div>
     </div>
 </div>

--- a/esp/src/src/Graph7Widget.ts
+++ b/esp/src/src/Graph7Widget.ts
@@ -1,0 +1,254 @@
+import * as lang from "dojo/_base/lang";
+import "dojo/i18n";
+// @ts-ignore
+import * as nlsHPCC from "dojo/i18n!hpcc/nls/hpcc";
+import * as aspect from "dojo/aspect";
+import * as dom from "dojo/dom";
+
+import * as registry from "dijit/registry";
+
+import { hashSum } from "@hpcc-js/util";
+import { ScopeGraph, Workunit } from "@hpcc-js/comms";
+import { Graph as GraphWidget, Subgraph, Vertex } from "@hpcc-js/graph";
+
+// @ts-ignore
+import * as _Widget from "hpcc/_Widget";
+
+// @ts-ignore
+import * as template from "dojo/text!hpcc/templates/Graph7Widget.html";
+
+import "dijit/layout/BorderContainer";
+import "dijit/layout/ContentPane";
+import "dijit/Toolbar";
+import "dijit/ToolbarSeparator";
+
+import declareDecorator from './DeclareDecorator';
+import { WUScopeController } from "./WUScopeController";
+
+type _Widget = {
+    id: string;
+    widget: any;
+    params: { [key: string]: any };
+    inherited(args: any);
+    setDisabled(id: string, disabled: boolean, icon?: string, disabledIcon?: string);
+};
+
+export interface Graph7Widget extends _Widget {
+}
+
+@declareDecorator("Graph7Widget", _Widget)
+export class Graph7Widget {
+    templateString = template;
+    static baseClass = "Graph7Widget";
+    i18n = nlsHPCC;
+
+    wuid = "";
+
+    graphStatus = null;
+
+    _graph: GraphWidget;
+    _gc = new WUScopeController();
+
+    constructor() {
+        this._gc.minClick = (sg: Subgraph) => {
+            this.loadGraph(w => {
+                this._graph
+                    .selection([sg])
+                    .centerOnItem(sg)
+                    ;
+            });
+        };
+    }
+
+    //  Data ---
+    private _prevHashSum;
+    private _prevScopeGraph: Promise<ScopeGraph>;
+    fetchScopeGraph(wuid: string, graphID: string, refresh: boolean = false): Promise<ScopeGraph> {
+        this.graphStatus.innerText = this.i18n.FetchingData;
+        const hash = hashSum({
+            wuid,
+            graphID
+        });
+        if (!this._prevScopeGraph || refresh || this._prevHashSum !== hash) {
+            this._prevHashSum = hash;
+            this._gc.clear();
+            const wu = Workunit.attach({ baseUrl: "" }, wuid);
+            this._prevScopeGraph = wu.fetchScopeGraphs(graphID ? [graphID] : []).then(scopedGraph => {
+                this.graphStatus.innerText = this.i18n.Loading;
+                return new Promise<ScopeGraph>((resolve, reject) => {
+                    setTimeout(() => {
+                        this._gc.set(scopedGraph);
+                        resolve(scopedGraph);
+                    }, 0);
+                });
+            });
+        }
+        return this._prevScopeGraph;
+    }
+    //  --- ---
+
+    buildRendering(args) {
+        this.inherited(arguments);
+    }
+
+    postCreate(args) {
+        this.inherited(arguments);
+        this._initGraphControls();
+    }
+
+    startup(args) {
+        this.inherited(arguments);
+    }
+
+    resize(s) {
+        this.inherited(arguments);
+        this.widget.MainBorderContainer.resize();
+    }
+
+    layout(args) {
+        this.inherited(arguments);
+    }
+
+    destroy(args) {
+        this.inherited(arguments);
+    }
+
+    //  Implementation  ---
+    _initGraphControls() {
+        aspect.after(registry.byId(this.id + "MainBorderContainer"), "resize", () => {
+            if (this._graph) {
+                this._graph
+                    .resize()
+                    .render()
+                    ;
+            }
+        });
+    }
+
+    _onRefresh() {
+        this.refreshData();
+    }
+
+    _onGraphRefresh() {
+        this._graph.data().subgraphs.forEach((sg: Subgraph) => {
+            sg.minState("normal");
+        });
+        delete this._graph["_prevLayout"];
+        this.loadGraph(w => {
+            this._graph.zoomToFit();
+        });
+    }
+
+    _onPartial(args) {
+        this._graph.data().subgraphs.forEach((sg: Subgraph) => {
+            sg.minState("partial");
+        });
+        this.loadGraph(w => {
+            this._graph.zoomToFit();
+        });
+    }
+
+    _onMax(args) {
+        this._graph.data().subgraphs.forEach((sg: Subgraph) => {
+            sg.minState("normal");
+        });
+        this.loadGraph(w => {
+            this._graph.zoomToFit();
+        });
+    }
+
+    _onZoomToFit(args) {
+        this._graph.zoomToFit();
+    }
+
+    _onZoomToWidth(args) {
+        this._graph.zoomToWidth();
+    }
+
+    _onZoomToPlus(args) {
+        this._graph.zoomPlus();
+    }
+
+    _onZoomToMinus(args) {
+        this._graph.zoomMinus();
+    }
+
+    isWorkunit() {
+        return lang.exists("params.Wuid", this);
+    }
+
+    isQuery() {
+        return lang.exists("params.QueryId", this);
+    }
+
+    init(params) {
+        if (this.inherited(arguments))
+            return;
+
+        this.initGraph();
+
+        this.doInit(params);
+    }
+
+    clear() {
+        this._graph
+            .data({ vertices: [], edges: [] })
+            .render()
+            ;
+    }
+
+    doInit(wuid: string) {
+        this.wuid = wuid;
+
+        this.refreshData();
+    }
+
+    refreshData() {
+        if (this.isWorkunit()) {
+            return this.loadGraphFromWu(this.wuid, "", true);
+        } else if (this.isQuery()) {
+        }
+        return Promise.resolve();
+    }
+
+    loadGraphFromWu(wuid, graphName, refresh: boolean = false) {
+        return this.fetchScopeGraph(wuid, graphName, refresh).then(() => {
+            this.loadGraph();
+        });
+    }
+
+    initGraph() {
+        this.graphStatus = dom.byId(this.id + "GraphStatus");
+        this._graph = new GraphWidget()
+            .target(this.id + "MainGraphWidget")
+            .layout("Hierarchy")
+            .applyScaleOnLayout(true)
+            .showToolbar(false)
+            .allowDragging(false)
+            .on("progress", what => {
+                switch (what) {
+                    case "start":
+                    case "layout-start":
+                    case "layout-tick":
+                        this.graphStatus.innerText = this.i18n.PerformingLayout;
+                        break;
+                    case "layout-end":
+                    case "end":
+                    default:
+                        this.graphStatus.innerText = "";
+                        break;
+                }
+            });
+        ;
+        this._graph.tooltipHTML((v: Vertex) => {
+            return this._gc.calcGraphTooltip2(v);
+        });
+    }
+
+    loadGraph(callback?) {
+        this._graph
+            .data(this._gc.graphData(), true)
+            .render(callback)
+            ;
+    }
+}

--- a/esp/src/src/GraphTree7Widget.ts
+++ b/esp/src/src/GraphTree7Widget.ts
@@ -105,7 +105,7 @@ export class GraphTree7Widget {
     private _prevHashSum;
     private _prevScopeGraph: Promise<ScopeGraph>;
     fetchScopeGraph(wuid: string, graphID: string, subgraphID: string = "", refresh: boolean = false): Promise<ScopeGraph> {
-        this.graphStatus.innerText = "Fetching...";
+        this.graphStatus.innerText = this.i18n.FetchingData;
         const hash = hashSum({
             wuid,
             graphID,
@@ -119,7 +119,7 @@ export class GraphTree7Widget {
                 for (const graph of graphs) {
                     if (graph.Name === graphID) {
                         return graph.fetchScopeGraph(subgraphID).then(scopedGraph => {
-                            this.graphStatus.innerText = "Loading...";
+                            this.graphStatus.innerText = this.i18n.Loading;
                             return new Promise<ScopeGraph>((resolve, reject) => {
                                 setTimeout(() => {
                                     this._gc.set(scopedGraph);
@@ -352,7 +352,18 @@ export class GraphTree7Widget {
                 this.syncSelectionFrom(this._graph);
             })
             .on("progress", what => {
-                this.graphStatus.innerText = what === "end" ? "" : what;
+                switch (what) {
+                    case "start":
+                    case "layout-start":
+                    case "layout-tick":
+                        this.graphStatus.innerText = this.i18n.PerformingLayout;
+                        break;
+                    case "layout-end":
+                    case "end":
+                    default:
+                        this.graphStatus.innerText = "";
+                        break;
+                }
             });
         ;
         this._graph.tooltipHTML((v: Vertex) => {

--- a/esp/src/src/Utility.ts
+++ b/esp/src/src/Utility.ts
@@ -465,7 +465,7 @@ export function resolve(hpccWidget, callback) {
             require(["src/GraphTree7Widget"], doLoad);
             break;
         case "Graph7Widget":
-            require(["hpcc/Graph7Widget"], doLoad);
+            require(["src/Graph7Widget"], doLoad);
             break;
         case "GridDetailsWidget":
             require(["hpcc/GridDetailsWidget"], doLoad);

